plugins/TwoFactorAuth/Controller.php

```
@@ -148,6 +152,13 @@ public function disableTwoFactorAuth()
             $this->twoFa->disable2FAforUser(Piwik::getCurrentUserLogin());
             $this->passwordVerify->forgetVerifiedPassword();
 
+            $container = StaticContainer::getContainer();
+            $email = $container->make(TwoFactorAuthDisabledEmail::class, array(
+                'login' => Piwik::getCurrentUserLogin(),
+                'emailAddress' => Piwik::getCurrentUserEmail()
+            ));
+            $email->send();
```

Just for completeness: If sending the mail fails for any reason this might throw an exception, which is currently uncaught.  It might then show an error message and the user might not be sure if the action was actually performed or not. Not sure if we should maybe better ignore possible exceptions in this case. Same applies for the other mails

---

I will double check this.

---

Now we have the abstract class, what if I add a method like this:
```php
public function safeSend()
{
    try {
        $this->send();
    } catch (\Exception $e) {
        // we do nothing if the email send was unsuccessful
    }
}
```

Any thoughts on this @sgiehl @diosmosis ?

---

It sounds like a good idea to not have exceptions from sending mail like these (which are side effects of other actions) break core functionality. I'm not sure we should just ignore the error though. Can we log it? eg,

```
catch (\Exception $e) {
    StaticContainer::get(LoggerInterface::class)->warning('Could not send whatever email: {exception}', ['exception' => $e]);
}
```